lib/charms/vault_k8s/v0/vault_client.py

```diff
@@ -515,68 +576,345 @@ def make_latest_pki_issuer_default(self, mount: str) -> None:
+
+
+class VaultAutounsealRequirerManager:
+    """Encapsulates the auto-unseal functionality from the Requirer Perspective.
+
+    In other words, this manages the feature from the perspective of the Vault
+    being auto-unsealed.
+    """
+
+    AUTOUNSEAL_TOKEN_SECRET_LABEL = "vault-autounseal-token"
+
+    def __init__(
+        self,
+        tls_manager: VaultTLSManager,
+        model: Model,
+        requires: VaultAutounsealRequires,
+    ):
+        self._tls_manager = tls_manager
+        self._model = model
+        self._requires = requires
+
+    def vault_configuration_details(self) -> AutounsealConfigurationDetails | None:
+        """Return the configuration details for the vault."""
+        autounseal_details = self._requires.get_details()
+        if not autounseal_details:
+            return None
+        self._tls_manager.push_autounseal_ca_cert(autounseal_details.ca_certificate)
+        ca_cert_path = self._tls_manager.get_tls_file_path_in_workload(File.AUTOUNSEAL_CA)
+        return AutounsealConfigurationDetails(
+            autounseal_details.address,
+            autounseal_details.mount_path,
+            autounseal_details.key_name,
+            self._get_autounseal_vault_token(autounseal_details),
+            ca_cert_path,
+        )
+
+    def _get_autounseal_vault_token(self, autounseal_details: AutounsealDetails) -> str:
+        """Retrieve the auto-unseal Vault token, or generate a new one if required.
+
+        Retrieves the last used token from Juju secrets, and validates that it
+        is still valid. If the token is not valid, a new token is generated and
+        stored in the Juju secret. A valid token is returned.
+
+        Args:
+            autounseal_details: The autounseal configuration details.
+
+        Returns:
+            A periodic Vault token that can be used for auto-unseal.
 
         """
-        key_name = self._create_autounseal_key(mount, relation_id)
-        policy_name = self._get_autounseal_policy_name(relation_id)
-        self.configure_policy(policy_name, policy_path, mount=mount, key_name=key_name)
+        external_vault = VaultClient(
+            url=autounseal_details.address,
+            ca_cert_path=self._tls_manager.get_tls_file_path_in_charm(File.AUTOUNSEAL_CA),
+        )
+        existing_token = self._get_juju_secret_field(self.AUTOUNSEAL_TOKEN_SECRET_LABEL, "token")
+        # If we don't already have a token, or if the existing token is invalid,
+        # authenticate with the AppRole details to generate a new token.
+        if not existing_token or not external_vault.authenticate(Token(existing_token)):
+            external_vault.authenticate(
+                AppRole(autounseal_details.role_id, autounseal_details.secret_id)
+            )
+            # NOTE: This is a little hacky. If the token expires, every unit
+            # will generate a new token, until the leader unit generates a new
+            # valid token and sets it in the Juju secret.
```

Comment:
It might be better to just put the charm in waiting status here...

Reply:
We probably don't want to have the charm's status management handled by the Vault API client

Reply:
@gruyaume Yeah, it would have to be done via a return value.

Reply:
If you have any preference, let me know. I left this comment to spark discussion if there is any preference here, but I went with the easiest option.